contextual.go

```diff
@@ -65,22 +65,54 @@ var (
 // To remove a backing logr implemention, use ClearLogger. Setting an
 // empty logger with SetLogger(logr.Logger{}) does not work.
 //
+// If the logger instance implements the ContextualLogger interface in addition
+// to logr.Logger, then the instance will also be called directly
+// by code that retrieves a logger with FromContext, TODO, or Background.
+// In that case the logger instance should do its own verbosity checks
+// in Enabled because klog will not do that for it.
+//
+// AddSupportsContext can be used for loggers which don't know about this
+// klog specific convention.
+//
 // Modifying the logger is not thread-safe and should be done while no other
 // goroutines invoke log calls, usually during program initialization.
 func SetLogger(logger logr.Logger) {
 	globalLogger = &logger
 	contextualLogger = false
+	if contextual, ok := logger.GetSink().(ContextualLogSink); ok && contextual.SupportsContext() {
+		contextualLogger = true
+	}
 }
 
-// SetContextualLogger does the same as SetLogger, but in addition the
-// logger may also get called directly by code that retrieves it
-// with FromContext, TODO or Background. The logger therefore must
-// implements its own verbosity checking.
-func SetContextualLogger(logger logr.Logger) {
-	globalLogger = &logger
-	contextualLogger = true
+// ContextualLogSink is an extension of the logr.LogSink interface that
+// indicates to SetLogger whether the logger may get called directly by
+// application code. Only the LogSink seen by SetLogger needs to implement
+// this, clones of it as they occur during logger.V, logger.WithName or
+// logger.WithValues don't need it.
+type ContextualLogSink interface {
```

Comment:
Do we need this to be public?

Reply:
Only if we expect some LogSinks to implement it directly. They'll need it for compile-time checks that they implement the interface.

I'm fine with saying that only wrapping is supported.

Reply:
I've pushed an update which doesn't have such an interface at all anymore.